Don't throw any exception when future is cancelled

i'm using [the pyright pycharm plugin with basedpyright](https://docs.basedpyright.com/latest/installation/ides/#pycharm). this exception occurred while i was rapidly typing, backspacing and saving with a file watcher for ruff active that reformats the file. i was trying to reproduce some of these strange intermittent issues i've been experiencing.

```
java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2510)
	at com.redhat.devtools.lsp4ij.features.codeAction.quickfix.LSPLazyCodeActions.cancel(LSPLazyCodeActions.java:192)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.lambda$toMap$1(LSPDiagnosticsForServer.java:113)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.toMap(LSPDiagnosticsForServer.java:108)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.update(LSPDiagnosticsForServer.java:70)
	at com.redhat.devtools.lsp4ij.OpenedDocument.updateDiagnostics(OpenedDocument.java:64)
	at com.redhat.devtools.lsp4ij.DocumentContentSynchronizer.lambda$refreshPullDiagnostic$8(DocumentContentSynchronizer.java:369)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
```

after this crash occurred, this nonsense diagnostic appeared and won't go away:

![Image](https://github.com/user-attachments/assets/4eb20722-72e2-4e99-a28b-1242c0b39c22)

this is quite a large file, about 11k lines so it usually takes a few seconds for the language server to analyze the file, so i think it might be a timing related issue, but i'm not sure.

lsp4ij v0.12.0
pycharm community 2024.3.5

let me know if you need any additional info. sorry this isn't a minimal repro, but i can't seem to figure out how to reproduce these issues reliably, so i hope the stack trace is enough

after attempting to reproduce this a few more times i encountered another crash that caused a similar result:

```
ExternalToolPass: 

com.intellij.diagnostic.PluginException: annotator: com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticAnnotator@44cd726b (class com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticAnnotator) [Plugin: com.redhat.devtools.lsp4ij]
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:23)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:90)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.processError(ExternalToolPass.java:272)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:245)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:235)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$1(ExternalToolPass.java:178)
	at com.intellij.openapi.application.ReadAction.lambda$run$1(ReadAction.java:54)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runReadAction(AnyThreadWriteThreadingSupport.kt:314)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runReadAction(AnyThreadWriteThreadingSupport.kt:262)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:864)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:66)
	at com.intellij.openapi.application.ReadAction.run(ReadAction.java:53)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$2(ExternalToolPass.java:175)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.lambda$runUnderDisposeAwareIndicator$15(BackgroundTaskUtil.java:371)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$1(CoreProgressManager.java:223)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.use(trace.kt:45)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:222)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$14(CoreProgressManager.java:674)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:749)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:705)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:673)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:203)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.runUnderDisposeAwareIndicator(BackgroundTaskUtil.java:366)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:172)
	at com.intellij.util.ui.update.Update.execute$suspendImpl(Update.kt:61)
	at com.intellij.util.ui.update.Update.execute(Update.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue.executeUpdates$suspendImpl(MergingUpdateQueue.kt:371)
	at com.intellij.util.ui.update.MergingUpdateQueue.executeUpdates(MergingUpdateQueue.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invokeSuspend(MergingUpdateQueue.kt:351)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invoke(MergingUpdateQueue.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invoke(MergingUpdateQueue.kt)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invokeSuspend(SingleAlarm.kt:394)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invoke(SingleAlarm.kt)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invoke(SingleAlarm.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:62)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:155)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1.invokeSuspend(SingleAlarm.kt:392)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:608)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:873)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:763)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:750)
Caused by: java.lang.IllegalArgumentException: Invalid range specified: (454821, 454790); 
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:288)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:283)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:279)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:42)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:31)
	at com.insyncwithfoo.pyright.shared.RangesKt.getOffsetRange(Ranges.kt:30)
	at com.insyncwithfoo.pyright.lsp4ij.DiagnosticFeature.createAnnotation(DiagnosticFeature.kt:46)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticAnnotator.createAnnotation(LSPDiagnosticAnnotator.java:103)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticAnnotator.doApply(LSPDiagnosticAnnotator.java:82)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticAnnotator.doApply(LSPDiagnosticAnnotator.java:41)
	at com.redhat.devtools.lsp4ij.features.AbstractLSPExternalAnnotator.apply(AbstractLSPExternalAnnotator.java:41)
	at com.intellij.codeInsight.daemon.impl.AnnotationHolderImpl.applyExternalAnnotatorWithContext(AnnotationHolderImpl.java:240)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:242)
	... 41 more

```

```
java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2510)
	at com.redhat.devtools.lsp4ij.features.codeAction.quickfix.LSPLazyCodeActions.cancel(LSPLazyCodeActions.java:192)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.lambda$toMap$1(LSPDiagnosticsForServer.java:113)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.toMap(LSPDiagnosticsForServer.java:108)
	at com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticsForServer.update(LSPDiagnosticsForServer.java:70)
	at com.redhat.devtools.lsp4ij.OpenedDocument.updateDiagnostics(OpenedDocument.java:64)

```

I have noticed this error few times but I have no idea why we have this error. I need to understand why we can have this error.

```
Caused by: java.lang.IllegalArgumentException: Invalid range specified: (454821, 454790); 
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:288)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:283)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:279)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:42)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:31)
	at com.insyncwithfoo.pyright.shared.RangesKt.getOffsetRange(Ranges.kt:30)
```

this error comes from the com.insyncwithfoo.pyright plugin, please report it.

--- a/src/main/java/com/redhat/devtools/lsp4ij/LanguageServerWrapper.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/LanguageServerWrapper.java
@@ -31,6 +31,7 @@ import com.redhat.devtools.lsp4ij.client.features.FileUriSupport;
 import com.redhat.devtools.lsp4ij.client.features.LSPClientFeatures;
 import com.redhat.devtools.lsp4ij.features.diagnostics.LSPDiagnosticUtils;
 import com.redhat.devtools.lsp4ij.features.files.operations.FileOperationsManager;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.ClientCapabilitiesFactory;
 import com.redhat.devtools.lsp4ij.internal.editor.EditorFeatureManager;
 import com.redhat.devtools.lsp4ij.internal.editor.EditorFeatureType;
@@ -573,7 +574,7 @@ public class LanguageServerWrapper implements Disposable {
             }
 
             if (this.initializeFuture != null) {
-                this.initializeFuture.cancel(true);
+                CancellationSupport.cancel(this.initializeFuture);
                 this.initializeFuture = null;
             }
 
@@ -654,9 +655,7 @@ public class LanguageServerWrapper implements Disposable {
             }
         }
 
-        if (serverFuture != null) {
-            serverFuture.cancel(true);
-        }
+        CancellationSupport.cancel(serverFuture);
 
         if (provider != null) {
             provider.stop();

--- a/src/main/java/com/redhat/devtools/lsp4ij/ServerMessageHandler.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/ServerMessageHandler.java
@@ -29,6 +29,7 @@ import com.intellij.openapi.ui.popup.LightweightWindowEvent;
 import com.redhat.devtools.lsp4ij.client.features.FileUriSupport;
 import com.redhat.devtools.lsp4ij.console.LSPConsoleToolWindowPanel;
 import com.redhat.devtools.lsp4ij.features.documentation.MarkdownConverter;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.StringUtils;
 import com.redhat.devtools.lsp4ij.server.definition.LanguageServerDefinition;
 import org.eclipse.lsp4j.*;
@@ -54,13 +55,13 @@ public class ServerMessageHandler {
      * Implements the LSP <a href="https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_logMessage">window/logMessage</a> specification.
      *
      * @param serverDefinition the language server definition.
-     * @param params  the message request parameters.
-     * @param project the project.
+     * @param params           the message request parameters.
+     * @param project          the project.
      */
     public static void logMessage(@NotNull LanguageServerDefinition serverDefinition,
                                   @NotNull MessageParams params,
                                   @NotNull Project project) {
-        LSPConsoleToolWindowPanel.showLog(serverDefinition, params, project );
+        LSPConsoleToolWindowPanel.showLog(serverDefinition, params, project);
     }
 
     private static Icon messageTypeToIcon(MessageType type) {
@@ -103,8 +104,8 @@ public class ServerMessageHandler {
      * Implements the LSP <a href="https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessageRequest">window/showMessageRequest</a> specification.
      *
      * @param languageServerName the language server name.
-     * @param params  the message request parameters.
-     * @param project the project.
+     * @param params             the message request parameters.
+     * @param project            the project.
      */
     public static CompletableFuture<MessageActionItem> showMessageRequest(@NotNull String languageServerName,
                                                                           @NotNull ShowMessageRequestParams params,
@@ -135,21 +136,17 @@ public class ServerMessageHandler {
                             }
                         });
                     }
-                    notification.whenExpired(()-> {
-                        if (!future.isDone()) {
-                            future.cancel(true);
-                        }
+                    notification.whenExpired(() -> {
+                        CancellationSupport.cancel(future);
                     });
 
                     Notifications.Bus.notify(notification, project);
-                    var balloon= notification.getBalloon();
+                    var balloon = notification.getBalloon();
                     if (balloon != null) {
                         balloon.addListener(new JBPopupListener() {
                             @Override
                             public void onClosed(@NotNull LightweightWindowEvent event) {
-                                if (!future.isDone()) {
-                                    future.cancel(true);
-                                }
+                                CancellationSupport.cancel(future);
                             }
                         });
                     }

--- a/src/main/java/com/redhat/devtools/lsp4ij/client/indexing/ProjectIndexingManager.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/client/indexing/ProjectIndexingManager.java
@@ -18,6 +18,7 @@ import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.psi.PsiFile;
 import com.redhat.devtools.lsp4ij.LanguageServersRegistry;
 import com.redhat.devtools.lsp4ij.client.ExecuteLSPFeatureStatus;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import org.eclipse.lsp4j.jsonrpc.CompletableFutures;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -77,9 +78,7 @@ public class ProjectIndexingManager implements Disposable {
 
     @Override
     public void dispose() {
-        if (waitForIndexingAllFuture != null) {
-            waitForIndexingAllFuture.cancel(true);
-        }
+        CancellationSupport.cancel(waitForIndexingAllFuture);
     }
 
     public static CompletableFuture<Void> waitForIndexingAll() {

--- a/src/main/java/com/redhat/devtools/lsp4ij/dap/DAPDebugProcess.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/dap/DAPDebugProcess.java
@@ -34,6 +34,7 @@ import com.redhat.devtools.lsp4ij.dap.client.DAPStackFrame;
 import com.redhat.devtools.lsp4ij.dap.client.DAPSuspendContext;
 import com.redhat.devtools.lsp4ij.dap.configurations.DAPCommandLineState;
 import com.redhat.devtools.lsp4ij.dap.descriptors.DebugAdapterDescriptor;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.CompletableFutures;
 import com.redhat.devtools.lsp4ij.internal.StringUtils;
 import com.redhat.devtools.lsp4ij.settings.ServerTrace;
@@ -227,12 +228,8 @@ public class DAPDebugProcess extends XDebugProcess {
                     .executeOnPooledThread(() -> {
                         try {
                             status = Status.STOPPING;
-                            if (serverReadyFuture != null && !CompletableFutures.isDoneNormally(serverReadyFuture)) {
-                                serverReadyFuture.cancel(true);
-                            }
-                            if (connectToServerFuture != null && !CompletableFutures.isDoneNormally(connectToServerFuture)) {
-                                connectToServerFuture.cancel(true);
-                            }
+                            CancellationSupport.cancel(serverReadyFuture);
+                            CancellationSupport.cancel(connectToServerFuture);
                             if (parentClient != null) {
                                 parentClient.terminate();
                             }

--- a/src/main/java/com/redhat/devtools/lsp4ij/dap/DAPServerReadyTracker.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/dap/DAPServerReadyTracker.java
@@ -115,7 +115,12 @@ public class DAPServerReadyTracker extends CompletableFuture<Void> implements Pr
     public boolean cancel(boolean mayInterruptIfRunning) {
         processHandler.removeProcessListener(this);
         processHandler.putUserData(SERVER_READY_TRACKER_KEY, null);
-        return super.cancel(mayInterruptIfRunning);
+        try {
+            return super.cancel(mayInterruptIfRunning);
+        } catch (Throwable e) {
+            // Ignore any error while cancelling
+            return false;
+        }
     }
 
     @NotNull

--- a/src/main/java/com/redhat/devtools/lsp4ij/dap/client/DAPClient.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/dap/client/DAPClient.java
@@ -24,6 +24,7 @@ import com.redhat.devtools.lsp4ij.dap.DebugMode;
 import com.redhat.devtools.lsp4ij.dap.TransportStreams;
 import com.redhat.devtools.lsp4ij.dap.breakpoints.DAPBreakpointProperties;
 import com.redhat.devtools.lsp4ij.dap.descriptors.DebugAdapterDescriptor;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.StringUtils;
 import com.redhat.devtools.lsp4ij.settings.ServerTrace;
 import org.eclipse.lsp4j.debug.*;
@@ -326,7 +327,8 @@ public class DAPClient implements IDebugProtocolClient, Disposable {
     @Override
     public void dispose() {
         if (debugProtocolFuture != null) {
-            debugProtocolFuture.cancel(true);
+            CancellationSupport.cancel(debugProtocolFuture);
+
             /*
              * If the debugProtocolFuture is running the message loop on the same thread we
              * are in currently, the cancel call will set the interrupt flag. That isn't

--- a/src/main/java/com/redhat/devtools/lsp4ij/features/AbstractLSPFeatureSupport.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/features/AbstractLSPFeatureSupport.java
@@ -110,9 +110,7 @@ public abstract class AbstractLSPFeatureSupport<Params, Result> {
         // Store the CancellationSupport in a local variable to prevent from NPE (very rare case)
         CancellationSupport cancellation = cancellationSupport;
         var future = this.future;
-        if (future != null && !future.isCancelled() && !future.isDone()) {
-            future.cancel(true);
-        }
+        CancellationSupport.cancel(future);
         this.future = null;
         if (cancellation != null) {
             cancellation.cancel();

--- a/src/main/java/com/redhat/devtools/lsp4ij/features/codeAction/quickfix/LSPLazyCodeActions.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/features/codeAction/quickfix/LSPLazyCodeActions.java
@@ -21,6 +21,7 @@ import com.redhat.devtools.lsp4ij.LanguageServerItem;
 import com.redhat.devtools.lsp4ij.features.codeAction.CodeActionData;
 import com.redhat.devtools.lsp4ij.features.codeAction.LSPLazyCodeActionIntentionAction;
 import com.redhat.devtools.lsp4ij.features.codeAction.LSPLazyCodeActionProvider;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.CompletableFutures;
 import org.eclipse.lsp4j.*;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
@@ -188,8 +189,6 @@ public class LSPLazyCodeActions implements LSPLazyCodeActionProvider {
      * Cancel if needed the LSP request textDocument/codeAction
      */
     public void cancel() {
-        if (lspCodeActionRequest != null && !lspCodeActionRequest.isDone()) {
-            lspCodeActionRequest.cancel(true);
-        }
+        CancellationSupport.cancel(lspCodeActionRequest);
     }
 }

--- a/src/main/java/com/redhat/devtools/lsp4ij/features/codeLens/LSPCodeLensProvider.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/features/codeLens/LSPCodeLensProvider.java
@@ -28,6 +28,7 @@ import com.redhat.devtools.lsp4ij.LanguageServerItem;
 import com.redhat.devtools.lsp4ij.client.ExecuteLSPFeatureStatus;
 import com.redhat.devtools.lsp4ij.client.features.LSPCodeLensFeature;
 import com.redhat.devtools.lsp4ij.client.indexing.ProjectIndexingManager;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import com.redhat.devtools.lsp4ij.internal.CompletableFutures;
 import com.redhat.devtools.lsp4ij.internal.PsiFileChangedException;
 import com.redhat.devtools.lsp4ij.internal.StringUtils;
@@ -143,7 +144,7 @@ public class LSPCodeLensProvider implements CodeVisionProvider<Void> {
                             try {
                                 waitUntilDone(visibleCodeLensToResolve, psiFile);
                             } catch (PsiFileChangedException e) {
-                                visibleCodeLensToResolve.cancel(true);
+                                CancellationSupport.cancel(visibleCodeLensToResolve);
                                 return CodeVisionState.NotReady.INSTANCE;
                             }
                         }

--- a/src/main/java/com/redhat/devtools/lsp4ij/installation/ServerInstallerBase.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/installation/ServerInstallerBase.java
@@ -16,6 +16,7 @@ import com.intellij.openapi.progress.ProgressManager;
 import com.intellij.openapi.progress.Task;
 import com.intellij.openapi.project.Project;
 import com.redhat.devtools.lsp4ij.LanguageServerBundle;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
@@ -137,9 +138,9 @@ public abstract class ServerInstallerBase implements ServerInstaller {
                 } catch (ProcessCanceledException e) {
                     //Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
                     //TODO delete block when minimum required version is 2024.2
-                    installFuture.cancel(true);
+                    CancellationSupport.cancel(installFuture);
                 } catch (CancellationException e) {
-                    installFuture.cancel(true);
+                    CancellationSupport.cancel(installFuture);
                 } catch (Throwable e) {
                     status = ServerInstallationStatus.NOT_INSTALLED;
                     installFuture.completeExceptionally(e);

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/CancellationSupport.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/CancellationSupport.java
@@ -37,10 +37,7 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.List;
-import java.util.concurrent.CancellationException;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.CompletionException;
-import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.*;
 import java.util.function.BiFunction;
 
 /**
@@ -105,9 +102,7 @@ public class CancellationSupport implements CancelChecker {
                                             @Nullable String featureName,
                                             boolean handleLanguageServerError) {
         if (cancelled) {
-            if (!future.isDone()) {
-                future.cancel(true);
-            }
+            CancellationSupport.cancel(future);
             throw new ProcessCanceledException();
         } else {
             // Add the future to the list of the futures to cancel (when CancellationSupport.cancel() is called)
@@ -225,9 +220,7 @@ public class CancellationSupport implements CancelChecker {
         }
         this.cancelled = true;
         for (CompletableFuture<?> futureToCancel : futuresToCancel) {
-            if (!futureToCancel.isDone()) {
-                futureToCancel.cancel(true);
-            }
+            CancellationSupport.cancel(futureToCancel);
         }
         futuresToCancel.clear();
     }
@@ -251,12 +244,28 @@ public class CancellationSupport implements CancelChecker {
             if (t instanceof CancellationException) {
                 if (to != null) {
                     for (var f : to) {
-                        f.cancel(true);
+                        cancel(f);
                     }
                 }
             }
             return null;
         });
+    }
+
+    /**
+     * Cancel the given future with any error.
+     *
+     * @param future the future to cancel.
+     */
+    public static void cancel(@Nullable Future<?> future) {
+        if (future != null && future.isDone()) {
+            try {
+                future.cancel(true);
+            } catch (Throwable e) {
+                // Ignore any error while cancelling the future.
+                LOGGER.warn("Unexpected error while cancelling future", e);
+            }
+        }
     }
 
 }

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/CompletableFutures.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/CompletableFutures.java
@@ -144,7 +144,7 @@ public class CompletableFutures {
                 future.get(25, TimeUnit.MILLISECONDS);
             } catch (TimeoutException ignore) {
                 long time = System.currentTimeMillis() - start;
-                if (timeout != null  && time > timeout) {
+                if (timeout != null && time > timeout) {
                     throw ignore;
                 }
                 if (file != null && !future.isDone() && System.currentTimeMillis() - start > 5000 &&
@@ -198,15 +198,10 @@ public class CompletableFutures {
                 } catch (
                         ProcessCanceledException e) {//Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
                     //TODO delete block when minimum required version is 2024.2
-                    if (!future.isCancelled()) {
-                        // Case when user click on cancel of progress Task.
-                        future.cancel(true);
-                    }
+                    // Case when user click on cancel of progress Task.
+                    CancellationSupport.cancel(future);
                 } catch (CancellationException e) {
-                    if (!future.isCancelled()) {
-                        // Case when user click on cancel of progress Task.
-                        future.cancel(true);
-                    }
+                    CancellationSupport.cancel(future);
                 } catch (Exception e) {
                     // Do nothing, error should be handled by the block code
                     // which consumes the future:
@@ -217,7 +212,7 @@ public class CompletableFutures {
     }
 
     public static CompletableFuture<Void> allOf(CompletableFuture<?>... cfs) {
-        var allOff =  CompletableFuture.allOf(cfs);
+        var allOff = CompletableFuture.allOf(cfs);
         CancellationSupport.forwardCancellation(allOff, cfs);
         return allOff;
     }

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/PromiseToCompletableFuture.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/PromiseToCompletableFuture.java
@@ -143,7 +143,8 @@ public class PromiseToCompletableFuture<R> extends CompletableFuture<R> {
                     try {
                         R result = code.apply(indicator);
                         return new ResultOrError<R>(result, null);
-                    } catch (ProcessCanceledException e) {//Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
+                    } catch (
+                            ProcessCanceledException e) {//Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
                         //TODO delete block when minimum required version is 2024.2
                         return new ResultOrError<R>(null, e);
                     } catch (CancellationException | IndexNotReadyException e) {
@@ -180,7 +181,12 @@ public class PromiseToCompletableFuture<R> extends CompletableFuture<R> {
                 nonBlockingReadActionPromise.cancel();
             }
         }
-        return super.cancel(mayInterruptIfRunning);
+        try {
+            return super.cancel(mayInterruptIfRunning);
+        } catch (Throwable e) {
+            // Ignore any error while cancelling
+            return false;
+        }
     }
 
 }

